### PR TITLE
Add AbstractAdmin::createNewInstance method

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

Ref: https://github.com/sonata-project/SonataAdminBundle/pull/6850#issuecomment-779792471

In 4.0, if a user wants to override the way in which their objects are created, they will have to override `ModelManagerInterface::getModelInstance` for each class. 

The code from `ModelManagerInterface::getModelInstance()` in our persistence bundles (which has been copied to `Instantiator::instantiate()`), is not persistence related.

So this PR adds a `AdminInterface::createNewInstance()` method which the user can override to use a factory, service or any other pattern to create their objects.

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added `AbstractAdmin::createNewInstance()` to allow the user to customize how to create their objects.
### Deprecated
- Deprecated `ModelManagerInterface::getModelInstance()` method in favor of `AbstractAdmin::createNewInstance()`.
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Is it needed to make it public ? Cant it be protected instead ?

And should we deprecate alterNewInstance in favor of createNewInstance then ?

**Comment by franmomu** [comment]
> Is it needed to make it public ? Cant it be protected instead ?
> 
> And should we deprecate alterNewInstance in favor of createNewInstance then ?

mmm not really, I'll change it.

File: tests/App/Admin/TranslatedAdmin.php [hunk 1]
```diff
@@ -18,10 +18,16 @@
 use Sonata\AdminBundle\Datagrid\ListMapper;
 use Sonata\AdminBundle\Form\FormMapper;
 use Sonata\AdminBundle\Show\ShowMapper;
+use Sonata\AdminBundle\Tests\App\Model\Translated;
 use Symfony\Component\Form\Extension\Core\Type\TextType;
 
 final class TranslatedAdmin extends AbstractAdmin
 {
+    public function createNewInstance(): object
```
Comment (VincentLanglet) [inline tests/App/Admin/TranslatedAdmin.php:26]:
```suggestion
    protected function createNewInstance(): object
```

File: tests/App/Admin/FooAdmin.php [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
 class FooAdmin extends AbstractAdmin
 {
-    public function getNewInstance()
+    public function createNewInstance(): object
```
Comment (VincentLanglet) [inline tests/App/Admin/FooAdmin.php:28]: protected

**Comment by franmomu** [comment]
> And should we deprecate alterNewInstance in favor of createNewInstance then ?

I'm not sure about this one to be honest, since it's called after `appendParentObject` I guess it's different.

**Review by VincentLanglet — DISMISSED** [review]

**Review by dmaicher — COMMENTED** [review]
LGTM but the commit message (and PR title) are a bit misleading since we are not adding anything to `AdminInterface`

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @franmomu

**Pull request merged**